### ISSUE TITLE
Thread Safety Analysis "guarded_by" attribute doesn't work for struct fields in C code

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#20777]
|  |  |
| --- | --- |
| Bugzilla Link | [20403](https://llvm.org/bz20403) |
| Version | trunk |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @AaronBallman,@Benvanik,@LebedevRI,@mikaelholmen,@nashidau,@zygoloid |

## Extended Description
```
$ cat demo.c
struct mutex {
        int x;
};

struct guarded_int {
        struct mutex mu;
        int value __attribute__((guarded_by(mu))); 
};

$ clang -c demo.c
demo.c:7:38: error: use of undeclared identifier 'mu'
        int value __attribute__((guarded_by(mu)));
                                            ^
1 error generated.
```

Clang compiles the code okay with "-x c++", so it seems to be a bug in how Thread Safety Analysis is wired into the C frontend.

**Comment by melver** [comment]
As mentioned on cfe-dev, we're interested in using ThreadSafetyAnalysis for the Linux kernel.

Since this issue has existed for some years, what is required to get this unstuck? Any estimate on the implementation complexity to get this supported?

(Unfortunately I'm not at all familiar with frontend internals, so I wouldn't know where to start -- otherwise I would have tried to put together a prototype, as that feels like the best way to discuss further.)

**Comment by aaronpuchert** [comment]
> Since this issue has existed for some years, what is required to get this unstuck? Any estimate on the implementation complexity to get this supported?

Best qualified to answer this is probably @AaronBallman, but I'll give my 2 cents.

> While we probably don't want to introduce an implicit 'this' pointer into C, I think we still should consider doing that. We need a solution that works in both C and C++ with the same syntax (because the structure may be in a header file that's included from both .c and .cpp compilands) and this is the most obvious solution for users to reach for.

That's probably the easiest solution, but it's still not easy. C code generally doesn't have any `CXXThisExpr`, so there might be weird bugs or assertions. But you could just figure that out as you go along. (Perhaps we want to introduce a separate AST node type, but that comes with its own problems.)

At some point we have to insert the members of the `struct` into the lookup scope, as it's done in C++, and emit some `MemberExpr` with an implicit `CXXThisExpr`. The code is already there but probably guarded by a language check. We don't want to generally allow this in C, but only in the context of attributes though. A good place to start might be `Parser::ParseAttributeArgsCommon` in [clang/lib/Parse/ParseDecl.cpp](../blob/main/clang/lib/Parse/ParseDecl.cpp). The calls to `ParseIdentifierLoc` and `ParseAssignmentExpression` are where the magic happens. So I'd say we have to add some code before those calls.

My understanding is that the attributes are parsed in a "complete type" context, i.e. lookup and semantic analysis is deferred until we've seen all members. That's likely the same in C and C++, because why would we maintain two different ways to do it? But I'm just guessing here.

> We should consider how to represent the case where a member of a struct is notionally guarded by acquisition of the a lock on the entire struct:
> 
> struct Lockable { int internal_mutex; int value [[clang::guarded_by(*this)]]; }; void lock(Lockable *p) [[clang::exclusive_lock_function(*p)]]; // ...
>
> Allowing the attributes to name the struct members isn't enough to model this; we need to allow the attributes to name the struct itself.

Such self-referential behavior is probably not too common, at least I can't remember seeing a record containing members guarded by the record itself. It seems preferable for numerous reasons to have the mutex/capability a separate member from what it guards. So I agree with @AaronBallman, we can likely go without explicit `this` or special syntax.

**Comment by tbaederr** [comment]
I've looked into this a bit and it's kind of a weird issue where the reason this is broken is super general (there is nothing like a `this` pointer in C), but the fix is really only needed in this one case (as far as I know).

I have locally added something like `__builtin_instance_member(membername)`, which the thread safety analysis then translates using `Ctx->SelfArg`, in other words, the same thing it does for `CXXThisPtr`.

Would this be an acceptable solution or is that still too general?

**Comment by AaronBallman** [comment]
> I've looked into this a bit and it's kind of a weird issue where the reason this is broken is super general (there is nothing like a `this` pointer in C), but the fix is really only needed in this one case (as far as I know).

Not the only case. The [-fbounds-safety](https://discourse.llvm.org/t/rfc-enforcing-bounds-safety-in-c-fbounds-safety/70854) proposal also wants to use field names as an attribute argument in C.

**Comment by aaronpuchert** [comment]
> I have locally added something like `__builtin_instance_member(membername)`, which the thread safety analysis then translates using `Ctx->SelfArg`, in other words, the same thing it does for `CXXThisPtr`.
> 
> Would this be an acceptable solution or is that still too general?

Deferring to others here, but to me it sounds reasonable.

> Edit: Actually, doesn't a new attribute `guarded_by_instance_ptr` or similar make the most sense?

So in the initial example, instead of `__attribute__((guarded_by(mu)))`, which is equivalent to `__attribute__((guarded_by(this->mu)))`, we would write `__attribute__((guarded_by_instance_ptr(mu)))`? It's probably not quite as powerful as introducing `this` through the backdoor like the builtin would do, because then you could write arbitrary expressions involving it. But it might be powerful enough.

**Comment by delcypher** [comment]
Should be fixed by #94216
